extmod/vfs_lfsx.c

```diff
@@ -98,17 +99,68 @@ STATIC void MP_VFS_LFSx(init_config)(MP_OBJ_VFS_LFSx * self, mp_obj_t bdev, size
     #endif
 }
 
-const char *MP_VFS_LFSx(make_path)(MP_OBJ_VFS_LFSx * self, mp_obj_t path_in) {
+// replacement for strtok, which does not modify the source string
+// but returns instead the length of the token
+static char *get_token(const char *path, char sep, size_t *len) {
+    static const char *ptr = ""; // safe initial state
+    const char *start;
+    if (path != NULL) {  // get initial start
+        ptr = (char *)path;
+    }
+    while (*ptr == sep && *ptr != '\0') { // scan for the start
+        ptr++;
+    }
+    if (*ptr == '\0') {  // End of the scan
+        ptr = ""; // safe parking
+        return NULL;
+    }
+    start = ptr;
+    while (*ptr != '\0' && *ptr != sep) {
+        ptr++;
+    }
+    *len = (ptr - start);
+    return (char *)start;
+}
+
+char *MP_VFS_LFSx(make_path)(MP_OBJ_VFS_LFSx * self, mp_obj_t path_in) {
+    char *cwd = vstr_null_terminated_str(&self->cur_dir);
+    char *new_path;
     const char *path = mp_obj_str_get_str(path_in);
-    if (path[0] != '/') {
-        size_t l = vstr_len(&self->cur_dir);
-        if (l > 0) {
-            vstr_add_str(&self->cur_dir, path);
-            path = vstr_null_terminated_str(&self->cur_dir);
-            self->cur_dir.len = l;
+
+    // Initiliaze new path
+    // If cwd was empty (unlikely) or path starts with /, start with "/"
+    // otherwise set if to cwd, but cut off a trailing '/', which is added by chdir
+    new_path = m_malloc(strlen(cwd) + strlen(path) + 1);  // worst case
+    if (path[0] == '/' || cwd[0] == '\0') {
+        strcpy(new_path, "/");
+    } else {
+        strcpy(new_path, cwd);
+        size_t l = strlen(new_path) - 1; // remove trailing /, if any
+        if (l > 0 && new_path[l] == '/') {
+            new_path[l] = '\0';
+        }
+    }
+
+    size_t len_token;
+    size_t len_path = strlen(new_path);
+    for (char *token = get_token(path, '/', &len_token); token != NULL;) {
+        if (len_token == 2 && token[0] == '.' && token[1] == '.') { // double dot, skip back in new_path
+            while (len_path > 0 && new_path[len_path] != '/') { // skip back w/o strrchr
+                len_path--;
+            }
+            new_path[len_path == 0 ? 1 : len_path] = '\0';
+        } else if (len_token != 1 || token[0] != '.') {
+            if (len_path > 1) { // not at the start
+                new_path[len_path++] = '/';
+            }
+            for (size_t i = 0; i < len_token; i++) { // avoid strncat and strncpy
+                new_path[len_path++] = token[i];
+            }
+            new_path[len_path] = '\0';
         }
+        token = get_token(NULL, '/', &len_token);
```

Comment:
you could move this into the for statement...

Reply:
I moved that in and out. Stylewise I prefer to have it in, even if the generated code should not differ. Moving that into the loop body was is an attempt for silence a travis build fail, which still says in 13410.2:
```
In file included from ../../extmod/vfs_lfs.c:120:
../../extmod/vfs_lfsx.c:146:58: error: variable 'token' used in loop condition not modified in loop body [-Werror,-Wfor-loop-analysis]
    for (char *token = get_token(path, '/', &len_token); token != NULL;) {
```
All left-over travis complaints are now about re-definition of the function get_token, which is strange. 
```
In file included from ../../extmod/vfs_lfs.c:120:0:
../../extmod/vfs_lfsx.c:104:14: error: redefinition of ‘get_token’
 static char *get_token(const char *path, char sep, size_t *len) {
              ^
In file included from ../../extmod/vfs_lfs.c:74:0:
../../extmod/vfs_lfsx.c:104:14: note: previous definition of ‘get_token’ was here
 static char *get_token(const char *path, char sep, size_t *len) {
              ^
../../py/mkrules.mk:63: recipe for target 'build-coverage/extmod/vfs_lfs.o' failed
```

Reply:
Looking at the errors, I believe that the "token error" is a side-effect of the get_token redefinition error, so once you solve that the token error will go away.

Reply:
But there is no re-definition, or at least I see none. get_token is defined only once.

Reply:
Well, if you read the error messages carefully, it says `In file included from ../../extmod/vfs_lfs.c:120` and `In file included from ../../extmod/vfs_lfs.c:74`, so lines 74 and 120 both include `../../extmod/vfs_lfsx.c` and that's why there's a redefinition. (I haven't looked at the source.)

Reply:
But get_token is in the included vfs_lfsx.c, and therefore the definition is the same. I could declare it in vfs_lfs.c again, like it is done for make_path. Would that resolve the travis complaints?

Reply:
Look at vfs_lfs.c carefully. It includes vfs_lfsx.c twice, once for LFS1 and once for LFS2. It defines some macros differently for each time so the functions end up being different and have different names. That's why there's the weird MP_VFS_LFSx(make_new) type of declarations. You need to move get_token out of the vfs_lfsx.c file.

Reply:
But get_token does not use any of these macros.

Reply:
Changed, and now it passes.